File: study/examples/Algorithm/cpp/05_sorting.cpp
/*
 * 정렬 알고리즘 (Sorting Algorithms)
 * Bubble, Selection, Insertion, Merge, Quick, Heap, Counting, Radix
 *
 * 다양한 정렬 알고리즘의 구현과 비교입니다.
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <cmath>

using namespace std;

// =============================================================================
// 1. 버블 정렬 - O(n²)
// =============================================================================

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }
        if (!swapped) break;  // 최적화
    }
}

// =============================================================================
// 2. 선택 정렬 - O(n²)
// =============================================================================

void selectionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        swap(arr[i], arr[minIdx]);
    }
}

// =============================================================================
// 3. 삽입 정렬 - O(n²)
// =============================================================================

void insertionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// =============================================================================
// 4. 병합 정렬 - O(n log n)
// =============================================================================

void merge(vector<int>& arr, int left, int mid, int right) {
    vector<int> L(arr.begin() + left, arr.begin() + mid + 1);
    vector<int> R(arr.begin() + mid + 1, arr.begin() + right + 1);

    size_t i = 0, j = 0;
    int k = left;

    while (i < L.size() && j < R.size()) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < L.size()) arr[k++] = L[i++];
    while (j < R.size()) arr[k++] = R[j++];
}

void mergeSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// =============================================================================
// 5. 퀵 정렬 - O(n log n) 평균
// =============================================================================

int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// 3-way 퀵소트 (중복 원소에 효율적)
void quickSort3Way(vector<int>& arr, int low, int high) {
    if (low >= high) return;

    int lt = low, gt = high;
    int pivot = arr[low];
    int i = low + 1;

    while (i <= gt) {
        if (arr[i] < pivot) {
            swap(arr[lt++], arr[i++]);
        } else if (arr[i] > pivot) {
            swap(arr[i], arr[gt--]);
        } else {
            i++;
        }
    }

    quickSort3Way(arr, low, lt - 1);
    quickSort3Way(arr, gt + 1, high);
}

// =============================================================================
// 6. 힙 정렬 - O(n log n)
// =============================================================================

void heapify(vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

void heapSort(vector<int>& arr) {
    int n = arr.size();

    // 힙 구성
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // 하나씩 추출
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

// =============================================================================
// 7. 계수 정렬 - O(n + k)
// =============================================================================

void countingSort(vector<int>& arr) {
    if (arr.empty()) return;

    int maxVal = *max_element(arr.begin(), arr.end());
    int minVal = *min_element(arr.begin(), arr.end());
    int range = maxVal - minVal + 1;

    vector<int> count(range, 0);
    vector<int> output(arr.size());

    for (int x : arr) {
        count[x - minVal]++;
    }

    for (int i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }

    for (int i = arr.size() - 1; i >= 0; i--) {
        output[count[arr[i] - minVal] - 1] = arr[i];
        count[arr[i] - minVal]--;
    }

    arr = output;
}

// =============================================================================
// 8. 기수 정렬 - O(d × (n + k))
// =============================================================================

void countingSortForRadix(vector<int>& arr, int exp) {
    int n = arr.size();
    vector<int> output(n);
    vector<int> count(10, 0);

    for (int x : arr) {
        count[(x / exp) % 10]++;
    }

    for (int i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    for (int i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }

    arr = output;
}

void radixSort(vector<int>& arr) {
    if (arr.empty()) return;

    int maxVal = *max_element(arr.begin(), arr.end());

    for (int exp = 1; maxVal / exp > 0; exp *= 10) {
        countingSortForRadix(arr, exp);
    }
}

// =============================================================================
// 테스트
// =============================================================================

void printVector(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        cout << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    cout << "============================================================" << endl;
    cout << "정렬 알고리즘 예제" << endl;
    cout << "============================================================" << endl;

    vector<int> original = {64, 34, 25, 12, 22, 11, 90};

    // 1. 버블 정렬
    cout << "\n[1] 버블 정렬" << endl;
    vector<int> arr1 = original;
    bubbleSort(arr1);
    cout << "    결과: ";
    printVector(arr1);
    cout << endl;

    // 2. 선택 정렬
    cout << "\n[2] 선택 정렬" << endl;
    vector<int> arr2 = original;
    selectionSort(arr2);
    cout << "    결과: ";
    printVector(arr2);
    cout << endl;

    // 3. 삽입 정렬
    cout << "\n[3] 삽입 정렬" << endl;
    vector<int> arr3 = original;
    insertionSort(arr3);
    cout << "    결과: ";
    printVector(arr3);
    cout << endl;

    // 4. 병합 정렬
    cout << "\n[4] 병합 정렬" << endl;
    vector<int> arr4 = original;
    mergeSort(arr4, 0, arr4.size() - 1);
    cout << "    결과: ";
    printVector(arr4);
    cout << endl;

    // 5. 퀵 정렬
    cout << "\n[5] 퀵 정렬" << endl;
    vector<int> arr5 = original;
    quickSort(arr5, 0, arr5.size() - 1);
    cout << "    결과: ";
    printVector(arr5);
    cout << endl;

    // 6. 힙 정렬
    cout << "\n[6] 힙 정렬" << endl;
    vector<int> arr6 = original;
    heapSort(arr6);
    cout << "    결과: ";
    printVector(arr6);
    cout << endl;

    // 7. 계수 정렬
    cout << "\n[7] 계수 정렬" << endl;
    vector<int> arr7 = original;
    countingSort(arr7);
    cout << "    결과: ";
    printVector(arr7);
    cout << endl;

    // 8. 기수 정렬
    cout << "\n[8] 기수 정렬" << endl;
    vector<int> arr8 = {170, 45, 75, 90, 802, 24, 2, 66};
    cout << "    원본: ";
    printVector(arr8);
    cout << endl;
    radixSort(arr8);
    cout << "    결과: ";
    printVector(arr8);
    cout << endl;

    // 9. 복잡도 비교
    cout << "\n[9] 정렬 알고리즘 복잡도 비교" << endl;
    cout << "    | 알고리즘   | 최선      | 평균      | 최악      | 공간  | 안정 |" << endl;
    cout << "    |------------|-----------|-----------|-----------|-------|------|" << endl;
    cout << "    | 버블       | O(n)      | O(n²)     | O(n²)     | O(1)  | O    |" << endl;
    cout << "    | 선택       | O(n²)     | O(n²)     | O(n²)     | O(1)  | X    |" << endl;
    cout << "    | 삽입       | O(n)      | O(n²)     | O(n²)     | O(1)  | O    |" << endl;
    cout << "    | 병합       | O(n log n)| O(n log n)| O(n log n)| O(n)  | O    |" << endl;
    cout << "    | 퀵         | O(n log n)| O(n log n)| O(n²)     | O(log)| X    |" << endl;
    cout << "    | 힙         | O(n log n)| O(n log n)| O(n log n)| O(1)  | X    |" << endl;
    cout << "    | 계수       | O(n+k)    | O(n+k)    | O(n+k)    | O(k)  | O    |" << endl;
    cout << "    | 기수       | O(d(n+k)) | O(d(n+k)) | O(d(n+k)) | O(n+k)| O    |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}

File: study/examples/Algorithm/python/05_sorting.py
"""
정렬 알고리즘 (Sorting Algorithms)
Sorting Algorithms Comparison

다양한 정렬 알고리즘의 구현과 비교입니다.
"""

import random
import time
from typing import List


# =============================================================================
# 1. 버블 정렬 (Bubble Sort)
# =============================================================================
def bubble_sort(arr: List[int]) -> List[int]:
    """
    버블 정렬
    시간: O(n²), 공간: O(1), 안정 정렬
    """
    arr = arr.copy()
    n = len(arr)

    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break

    return arr


# =============================================================================
# 2. 선택 정렬 (Selection Sort)
# =============================================================================
def selection_sort(arr: List[int]) -> List[int]:
    """
    선택 정렬
    시간: O(n²), 공간: O(1), 불안정 정렬
    """
    arr = arr.copy()
    n = len(arr)

    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

    return arr


# =============================================================================
# 3. 삽입 정렬 (Insertion Sort)
# =============================================================================
def insertion_sort(arr: List[int]) -> List[int]:
    """
    삽입 정렬
    시간: O(n²), 최선 O(n), 공간: O(1), 안정 정렬
    거의 정렬된 데이터에 효율적
    """
    arr = arr.copy()
    n = len(arr)

    for i in range(1, n):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

    return arr


# =============================================================================
# 4. 병합 정렬 (Merge Sort)
# =============================================================================
def merge_sort(arr: List[int]) -> List[int]:
    """
    병합 정렬
    시간: O(n log n), 공간: O(n), 안정 정렬
    """
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


def merge(left: List[int], right: List[int]) -> List[int]:
    """두 정렬된 배열 병합"""
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result


# =============================================================================
# 5. 퀵 정렬 (Quick Sort)
# =============================================================================
def quick_sort(arr: List[int]) -> List[int]:
    """
    퀵 정렬
    시간: 평균 O(n log n), 최악 O(n²), 공간: O(log n), 불안정 정렬
    """
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort(left) + middle + quick_sort(right)


def quick_sort_inplace(arr: List[int], low: int = 0, high: int = None) -> List[int]:
    """퀵 정렬 (in-place)"""
    if high is None:
        arr = arr.copy()
        high = len(arr) - 1

    if low < high:
        pivot_idx = partition(arr, low, high)
        quick_sort_inplace(arr, low, pivot_idx - 1)
        quick_sort_inplace(arr, pivot_idx + 1, high)

    return arr


def partition(arr: List[int], low: int, high: int) -> int:
    """파티션 (Lomuto scheme)"""
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


# =============================================================================
# 6. 힙 정렬 (Heap Sort)
# =============================================================================
def heap_sort(arr: List[int]) -> List[int]:
    """
    힙 정렬
    시간: O(n log n), 공간: O(1), 불안정 정렬
    """
    arr = arr.copy()
    n = len(arr)

    # 최대 힙 구성
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # 하나씩 추출
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

    return arr


def heapify(arr: List[int], n: int, i: int):
    """최대 힙 속성 유지"""
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


# =============================================================================
# 7. 계수 정렬 (Counting Sort)
# =============================================================================
def counting_sort(arr: List[int]) -> List[int]:
    """
    계수 정렬
    시간: O(n + k), 공간: O(k), 안정 정렬
    k = 최댓값 - 최솟값 + 1
    정수 범위가 작을 때 효율적
    """
    if not arr:
        return []

    min_val, max_val = min(arr), max(arr)
    range_val = max_val - min_val + 1

    count = [0] * range_val
    output = [0] * len(arr)

    # 카운트
    for num in arr:
        count[num - min_val] += 1

    # 누적
    for i in range(1, range_val):
        count[i] += count[i - 1]

    # 역순으로 배치 (안정성 유지)
    for i in range(len(arr) - 1, -1, -1):
        num = arr[i]
        count[num - min_val] -= 1
        output[count[num - min_val]] = num

    return output


# =============================================================================
# 8. 기수 정렬 (Radix Sort)
# =============================================================================
def radix_sort(arr: List[int]) -> List[int]:
    """
    기수 정렬 (LSD)
    시간: O(d * (n + k)), 공간: O(n + k)
    d = 자릿수, k = 기수 (보통 10)
    음수 미지원 버전
    """
    if not arr or min(arr) < 0:
        return sorted(arr)  # 음수 포함 시 기본 정렬

    max_val = max(arr)

    exp = 1
    while max_val // exp > 0:
        arr = counting_sort_by_digit(arr, exp)
        exp *= 10

    return arr


def counting_sort_by_digit(arr: List[int], exp: int) -> List[int]:
    """특정 자릿수 기준 계수 정렬"""
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for num in arr:
        digit = (num // exp) % 10
        count[digit] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        digit = (arr[i] // exp) % 10
        count[digit] -= 1
        output[count[digit]] = arr[i]

    return output


# =============================================================================
# 성능 비교
# =============================================================================
def benchmark_sorts(size: int = 1000):
    """정렬 알고리즘 성능 비교"""
    arr = [random.randint(0, 10000) for _ in range(size)]

    algorithms = [
        ("Bubble Sort", bubble_sort, size <= 1000),
        ("Selection Sort", selection_sort, size <= 1000),
        ("Insertion Sort", insertion_sort, size <= 1000),
        ("Merge Sort", merge_sort, True),
        ("Quick Sort", quick_sort, True),
        ("Heap Sort", heap_sort, True),
        ("Counting Sort", counting_sort, True),
        ("Radix Sort", radix_sort, True),
        ("Python sorted()", sorted, True),
    ]

    print(f"\n배열 크기: {size}")
    print("-" * 50)

    for name, func, should_run in algorithms:
        if should_run:
            test_arr = arr.copy()
            start = time.perf_counter()
            result = func(test_arr)
            elapsed = time.perf_counter() - start
            print(f"{name:20s}: {elapsed * 1000:8.3f} ms")
        else:
            print(f"{name:20s}: (크기 초과로 스킵)")


# =============================================================================
# 테스트
# =============================================================================
def main():
    print("=" * 60)
    print("정렬 알고리즘 (Sorting Algorithms)")
    print("=" * 60)

    # 테스트 배열
    arr = [64, 34, 25, 12, 22, 11, 90]
    print(f"\n원본 배열: {arr}")
    print("-" * 40)

    # 각 알고리즘 테스트
    algorithms = [
        ("버블 정렬", bubble_sort),
        ("선택 정렬", selection_sort),
        ("삽입 정렬", insertion_sort),
        ("병합 정렬", merge_sort),
        ("퀵 정렬", quick_sort),
        ("힙 정렬", heap_sort),
        ("계수 정렬", counting_sort),
        ("기수 정렬", radix_sort),
    ]

    for name, func in algorithms:
        result = func(arr.copy())
        print(f"{name}: {result}")

    # 성능 비교
    print("\n" + "=" * 60)
    print("성능 비교 (Performance Benchmark)")
    print("=" * 60)

    for size in [100, 1000, 5000]:
        benchmark_sorts(size)

    # 정렬 알고리즘 비교표
    print("\n" + "=" * 60)
    print("정렬 알고리즘 비교")
    print("=" * 60)
    print("""
    | 알고리즘     | 평균      | 최악      | 공간   | 안정성 | 특징                |
    |-------------|----------|----------|--------|-------|---------------------|
    | 버블 정렬    | O(n²)    | O(n²)    | O(1)   | 안정  | 단순, 교육용         |
    | 선택 정렬    | O(n²)    | O(n²)    | O(1)   | 불안정| 단순, 교환 횟수 적음  |
    | 삽입 정렬    | O(n²)    | O(n²)    | O(1)   | 안정  | 거의 정렬된 데이터에 좋음|
    | 병합 정렬    | O(nlogn) | O(nlogn) | O(n)   | 안정  | 일정한 성능          |
    | 퀵 정렬      | O(nlogn) | O(n²)    | O(logn)| 불안정| 평균적으로 가장 빠름  |
    | 힙 정렬      | O(nlogn) | O(nlogn) | O(1)   | 불안정| in-place, 일정한 성능|
    | 계수 정렬    | O(n+k)   | O(n+k)   | O(k)   | 안정  | 정수, 범위 작을 때    |
    | 기수 정렬    | O(d(n+k))| O(d(n+k))| O(n+k) | 안정  | 자릿수 기반           |

    실무 선택 가이드:
    - 일반적인 경우: 퀵 정렬 또는 언어 내장 정렬
    - 안정성 필요: 병합 정렬
    - 메모리 제한: 힙 정렬
    - 거의 정렬됨: 삽입 정렬
    - 정수 + 범위 작음: 계수/기수 정렬
    """)


if __name__ == "__main__":
    main()
